### PR TITLE
k256: Taproot Schnorr (BIP 340)

### PR DESCRIPTION
**Pull request by nickray**
Implements Taproot-flavored Schnorr signatures as described in:
https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki

Based on https://github.com/RustCrypto/elliptic-curves/pull/482, rebased onto master branch.

### CODE REVIEW

File: k256/src/schnorr.rs [hunk 1]
```diff
@@ -0,0 +1,518 @@
...
+///
+/// [BIP340]: https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki
+#[derive(Copy, Clone, Debug, Eq, PartialEq, PartialOrd, Ord)]
+pub struct Signature {
+    // TODO(nickray): maybe store (FieldElement, NonScalar), avoiding the unwraps.
+    // However, this
+    // a) contradicts tarcieri's philosophy of storing "bag of bytes"
+    // b) makes the Debug, Eq, PartialEq, PartialOrd, Ord impls tricky,
+    // as FieldElement + NonZeroScalar don't implement them, so we'd have to
+    // either modify upstream or implement in terms of the calculated `to_bytes`.
```
Comment (tarcieri) [inline k256/src/schnorr.rs:35]: I think it's fine to store the parsed components of the signature alongside `bytes`.

To simplify some of the questions you can remove the `PartialOrd` and `Ord` derives. I'm not sure they're useful for anything.

A handwritten `Debug` would probably make sense, either displaying the byte serialization or the parsed components of the signature.

File: k256/src/schnorr.rs [hunk 2]
```diff
@@ -0,0 +1,518 @@
...
+impl Signature {
+    /// Parse signature from big endian-encoded bytes.
+    pub fn from_bytes(bytes: &[u8]) -> Result<Self> {
+        let bytes: [u8; 64] = bytes.try_into()
+            .map_err(|_| Error::new())?;
+
+        let _: FieldElement = Option::from(FieldElement::from_bytes(FieldBytes::from_slice(&bytes[..32])))
+            .ok_or_else(Error::new)?;
+        let _: NonZeroScalar = Option::from(NonZeroScalar::from_repr(*FieldBytes::from_slice(&bytes[32..])))
+            .ok_or_else(Error::new)?;
```
Comment (tarcieri) [inline k256/src/schnorr.rs:48]:
```suggestion
    NonZeroScalar::try_from(&bytes[32..])?;
```

File: k256/src/arithmetic/affine.rs [hunk 3]
```diff
@@ -186,16 +197,22 @@ impl DecompressPoint<Secp256k1> for AffinePoint {
...
         })
     }
 }
 
+/// Decompaction using Taproot conventions as described in [BIP 340].
+///
+/// [BIP 340]: https://github.com/bitcoin/bips/blob/master/bip-0340.mediawiki
+impl DecompactPoint<Secp256k1> for AffinePoint {
+    fn decompact(x_bytes: &FieldBytes) -> CtOption<Self> {
+        // TODO(tarcieri): implement full `lift_x` algorithm as described in BIP 340
```
Comment (tarcieri) [inline k256/src/arithmetic/affine.rs:211]: Still an important TODO here

Comment (nickray) [reply k256/src/arithmetic/affine.rs:211]: As mentioned in the other PR, I'm missing what the remaining TODO is, so removing the marker. FieldElement::sqrt calculates y via the (p+1)/4'th power, checking it's an actual square root.

**Comment by tarcieri** [comment]
Will go ahead and merge this.

There are a few things I think can be addressed in a followup PR, most notably implementing the `signature` crate's traits. I'll try to take a look at that in the next few days.

**Pull request merged**